### ISSUE TITLE
IPLD preparations

### ISSUE THREAD

**Issue by yusefnapora** [mediachain/mediachain#158]
I did some reading through the updated [ipld spec](https://github.com/ipfs/specs/tree/master/ipld) this morning, as well as the [js-ipld](https://github.com/ipfs/js-ipld) implementation.

The biggest breaking change from our perspective is that the `@link` tag in the json representation has been replaced with `/`, and "link objects" are now only allowed to have a single key.  So a link would now look like e.g.:

``` json
{"artist": {"/": "QmF00..."}}
```

Link properties are now siblings of the "link object" instead of being contained inside it, e.g.:

``` json
{
  "thing_with_properties": {
    "prop": "value",
    "link": {"/": "QmF00..."}
  }
}
```

Since we're not using link properties that seems like it shouldn't affect us, but i'm a bit confused by [Link Properties Convention section](https://github.com/ipfs/specs/tree/master/ipld#link-properties-convention).  It's not clear how to distinguish "pseudo link objects" from plain objects that happen to contain links, which has implications for path resolution.

For example, if we have:

``` js
{
  type: "artefact",
  meta: {
    data: {
      title: "Hello world",
      artist: {"/": "Qm123"}
    }
  }
}
```

Would the path `/ipfs/<hash-of-that-object>/meta/data` just assume that we were trying to attach link properties to the `artist` link and replace the value of `data` with the contents of the artist link?  It would be good to get clarification on how path resolution works with these "pseudo link objects" before we dive in.  The examples all use the key `link` for the "real link", but it's not at all clear whether that key is "special" when resolving paths.

The canonical cbor representation hasn't changed much, but it does require the use of cbor tags, which we aren't yet using.  The spec still lists "register tag with IANA" as a TODO, but in the js implementation they're [using 258 as the tag](https://github.com/ipfs/js-ipld/blob/master/src/cbor.js#L13), so it's probably safe to use that.  Our cbor serialization code will need to be updated to use link tags instead of writing out cbor maps.

So, here's some things to do:
- [x] resolve ambiguity about link properties and path resolution
- [ ] update cbor serialization code to use tags instead of link maps  (needs client-side support as well)
- [ ] experiment with js-ipld implementation
  - see if we can put our raw cbor bytes into ipfs and resolve with [js-ipld-cli](https://github.com/ipfs/js-ipld-cli)

**Comment by jbenet** [comment]
Hey! Going to be super brief because of time atm, but i can expand later. cc also @nicola @diasdavid @dignifiedquire @mildred @davidar who can expand on some of this.
- on the "pesudo link objects". that's just a way to show you how one might create objects with more information. 
  - it's not part of the implementation or spec, really, it's just a "how do i add data to the links? oh, by making a wrapper object with the data there, not in the link itself, got it". 
  - Path resolution is super simple now (simpler than it used to be).
  - > Would the path /ipfs/<hash-of-that-object>/meta/data just assume that we were trying to attach link properties to the artist link and replace the value of data with the contents of the artist link?
    
    No, it would just be:
    
    ```
    > ipld resolve --encoding=json /ipfs/<hash-of-that-object>/meta/data
    {
      title: "Hello world",
      artist: {"/": "Qm123..."}
    }
    
    > ipld resolve --encoding=json /ipfs/<hash-of-that-object>/meta/data/artist
    <the object pointed to by Qm123...>
    ```
- > update cbor serialization code to use tags instead of link maps (needs client-side support as well)
  - @dignifiedquire @nicola @mildred @davidar o/ you understand the mapping to cbor better than i do. can you confirm the above?

**Comment by yusefnapora** [comment]
Awesome, thanks @jbenet!

**Comment by dignifiedquire** [comment]
> update cbor serialization code to use tags instead of link maps (needs client-side support as well)

I'm not a 100% sure what you mean by this. The changes that were made recently to the cbor tag encoding are the following:

### Before
- a text string (major type 3) or byte string (major type 2) corresponding to the link target. This is the canonical format for links with no link properties.
- an array (major type 4) containing as first element the link target (text or binary string) and as optional second argument the link properties (a map, major type 5)

### After

Only: a text string (major type 3) or byte string (major type 2) corresponding to the link target. This is the canonical format for links with no link properties.

For more details see this PR: https://github.com/ipfs/specs/pull/110/files which does exactly these changes to the spec.

**Comment by yusefnapora** [comment]
sweet, thanks for the explanation.  We just recently started working on using the correct canonical encoding for links (using cbor tags); so far we've just been writing out cbor maps of the form `{"@link": <multihash-bytes>}`, which isn't valid IPLD but got us close while the spec was in-progress.

That's what I was referring to about updating the serialization code to use tagged values instead of link maps.  

Looks like we're pretty close; #162 produces cbor objects with links that resolve using the js-ipld-cli.  At the moment we're writing out the multihashes as strings, but will probably use multiaddr soon.

**Comment by dignifiedquire** [comment]
Sounds good. Feel free to ping me when anything is unclear or is missing :)

**Comment by parkan** [comment]
Update from IPFS team:
- go implementation out of sync with standard but has some valuable optimizations (stream decoding)
- JS implementation = reference, basically there, can probably port it back
- waiting on timeline details so we can jump in to help
